Event system guide (#9328)
ckeditor/ckeditor5: packages/ckeditor5-engine/docs/framework/guides/deep-dive/event-system.md

---
category: framework-deep-dive
---

# Event system

{@link module:utils/emittermixin~Emitter Emitters} are objects that can fire events. They also provide means to bind into other emitters' events.

Emitters are heavily used in the {@link framework/guides/architecture/editing-engine#conversion conversion}, {@link framework/guides/architecture/editing-engine#observers Observers} and are a base for the {@link framework/guides/deep-dive/observables Observables}.  

Any class can become an event emitter; all you need to do is mix the {@link module:utils/emittermixin~EmitterMixin} into it:

```js
import EmitterMixin from '@ckeditor/ckeditor5-utils/src/emittermixin';
import mix from '@ckeditor/ckeditor5-utils/src/mix';

class AnyClass {
	// ...
}

mix( AnyClass, EmitterMixin );
```

<info-box>
	Check out the {@link framework/guides/deep-dive/custom-element-conversion custom element conversion} guide to learn more about the advanced usage of events in the conversion pipelines.
</info-box>

## Firing events

Once you mix the {@link module:utils/emittermixin~EmitterMixin} into your class you can fire events in the following way:

```js

> this.fire( 'eventName', ...args );

Author: This is a language standard so I would expect that it's clear

}
```

Note that {@link module:utils/eventinfo~EventInfo} expects source object in the first parameter as an origin of the event.

### Event return value

Listeners can set the {@link module:utils/eventinfo~EventInfo#return `eventInfo.return`} field, so the value is returned from the {@link module:utils/emittermixin~Emitter#fire `fire()`} call. 

```js
emitter.on( 'eventName', ( eventInfo, ...args ) => {
	eventInfo.return = 123;
} );

const result = emitter.fire( 'eventName', ...args );

console.log( result ); // -> 123
```

### Event namespaces

Event system supports namespaced events, so you could fire:

```js
this.fire( 'foo:bar:eventName', ...args );
```

Then listeners can be bound to a specific event or the whole namespace:

```js
this.on( 'foo', () => { ... } );
this.on( 'foo:bar', () => { ... } );
this.on( 'foo:bar:eventName', () => { ... } );
```

**Note**: Listeners registered on the same priority will be fired in the order of the registration (no matter if listening to a whole namespace or to the specific event).

## Listening to events

You could attach a listener directly on the emitter object:

```js
emitter.on( 'eventName', ( eventInfo, ...args ) => { ... } );
```

> but this way you would need to keep the reference to the handler function to be able to unregister that listener:

Reviewer: ```suggestion
However, a function object will be needed if you would need to be able to remove the event listener:
```

```js
emitter.off( 'eventName', handler );
```

> There is an easier way, you could use {@link module:utils/emittermixin~Emitter#listenTo `listenTo()`} method from one emitter and bind listener to the other emitter:

Reviewer: ```suggestion
There is also another way to add an event listener - by using {@link module:utils/emittermixin~Emitter#listenTo `listenTo()`}. This way one emitter can listen to events on another emitter:
```

```js
foo.listenTo( bar, 'eventName', ( eventInfo, ...args ) => { ... } );
```

> This way you could easily detach the `foo` from `bar` simply by {@link module:utils/emittermixin~Emitter#stopListening `stopListening()`}.  

Reviewer: ```suggestion
Now you can easily detach the `foo` from `bar` simply by {@link module:utils/emittermixin~Emitter#stopListening `stopListening()`}.  
```

```js
// Stop listening to a specific handler.
foo.stopListening( bar, 'eventName', handler );

// Stop listening to a specific event.
foo.stopListening( bar, 'eventName' );

// Stop listening to all events fired by a specific emitter.
foo.stopListening( bar );

// Stop listening to all events fired by all bound emitters.
foo.stopListening();
```

> **Note**: The above detaches only the listeners that were attached from `foo` on `bar`, all other listeners are not affected.

Reviewer: I don't think this note is needed, also it is not true for the last example.

<info-box>

>     The {@link module:utils/emittermixin~Emitter#on `on()`} and {@link module:utils/emittermixin~Emitter#off `off()`} methods are shorthands for {@link module:utils/emittermixin~Emitter#listenTo `listenTo( this, ... )`} and {@link module:utils/emittermixin~Emitter#stopListening `stopListening( this, ... )`} (emitter is bound on itself). 

Reviewer: ```suggestion
    The {@link module:utils/emittermixin~Emitter#on `on()`} and {@link module:utils/emittermixin~Emitter#off `off()`} methods are shorthands for {@link module:utils/emittermixin~Emitter#listenTo `listenTo( this, ... )`} and {@link module:utils/emittermixin~Emitter#stopListening `stopListening( this, ... )`} (emitter is bound to itself). 
```

Reviewer: But I think that this note is not needed as well.

It is possible to use relative priorities {@link module:utils/priorities~priorities#get `priorities.get( 'high' ) + 10`} but this is strongly discouraged.

## Bubbling events

The {@link module:engine/view/document~Document `view.Document`} is not only the {@link module:utils/observablemixin~Observable Observable} (and {@link module:utils/emittermixin~Emitter Emitter}) but it also implements {@link module:engine/view/observer/bubblingemittermixin~BubblingEmitter} interface. This is the special interface that is implemented by the {@link module:engine/view/observer/bubblingemittermixin~BubblingEmitterMixin}. It provides a bubbling of the events over the virtual DOM tree. It is different from the bubbling that you know from the browser's DOM tree events bubbling. You don't register listeners on the exact instances of the elements in the view document tree, instead you can register handlers for the `context` (for example the {@link module:engine/view/element~Element view element} name, the virtual `'$capture'`, `'$text'`, `'$root'`, `'$document'` contexts, or by providing a callback that matches some view nodes).  

> ### Listening to events

Reviewer: ```suggestion
### Listening to bubbling events
```
